core/includes/backup/backup.database.inc

```diff
@@ -0,0 +1,350 @@
+      }
+    }
+
+    $settings += array(
+      'nodata_tables' => backdrop_map_assoc(array_merge($basic, $cache)),
+      'exclude_tables' => $simpletest,
+      'lock_tables' => TRUE,
+      'compression' => self::COMPRESSION_GZIP,
+    );
+
+    return $settings;
+  }
+
+  /**
+   * Get the form for the backup settings for this destination.
+   */
+  public function backupSettingsForm() {
+    $objects = $this->getObjectNames();
+    $settings = $this->settings;
+
+    $form['#description'] = t('You may omit specific tables, or specific table data from the backup file. Only omit data that you know you will not need such as cache data, or tables from other applications. Excluding tables can break your Backdrop install, so <strong>do not change these settings unless you know what you\'re doing</strong>.');
+    $form['exclude_tables'] = array(
+      '#type' => 'select',
+      '#multiple' => TRUE,
+      '#title' => t('Exclude the following tables altogether'),
+      '#options' => $objects,
+      '#default_value' => $settings['exclude_tables'],
+      '#description' => t('The selected tables will not be added to the backup file.'),
+    );
+    $form['nodata_tables'] = array(
+      '#type' => 'select',
+      '#multiple' => TRUE,
+      '#title' => t('Exclude the data from the following tables'),
+      '#options' => $objects,
+      '#default_value' => $settings['nodata_tables'],
+      '#description' => t('The selected tables will have their structure backed up but not their contents. This is useful for excluding cache data to reduce file size.'),
+    );
+    $form['lock_tables'] = array(
+      '#type' => 'checkbox',
+      '#title' => t('Lock tables during backup'),
+      '#default_value' => !empty($settings['lock_tables']) ? $settings['lock_tables'] : NULL,
+      '#description' => t('This can help reduce data corruption, but will make your site unresponsive.'),
+    );
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function backup(BackupFile $file) {
+    $file->pushExtension($this->getFileTypeId());
+
+    if ($this->settings['lock_tables']) {
+      $this->lockTables();
+    }
+
+    // Switch to a different db if specified.
+    $success = $this->backupDatabaseToFile($file);
+
+    if ($this->settings['lock_tables']) {
+      $this->unlockTables();
+    }
+
+    // Pull any logged messages out of the file and up to this class.
+    $file_log = $file->getLog();
+    foreach ($file_log as $log) {
+      call_user_func_array($this->log, $log);
+    }
+    $file->clearLog();
+
+    return $success ? $file : FALSE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function postBackup(BackupFile &$file) {
+    parent::postBackup($file);
+    // Compress the backup and delete the original once compressed.
+    if ($this->settings['compression'] === self::COMPRESSION_GZIP) {
+      // The returned value should overwrite $file by reference.
+      $file = $this->compress($file, self::COMPRESSION_GZIP, TRUE);
```

Comment:
What happens, if compression fails? Could it fail?

Reply:
It's possible. Maybe if you run out of disk space or zlib isn't available (though you'd have other errors in Backdrop if that were the case). I'll add a graceful failure here.